File: app/src/main/java/com/dtu/capstone2/ereading/ui/newfeed/translate/TranslateNewFeedViewModel.java
package com.dtu.capstone2.ereading.ui.newfeed.translate;

import android.text.SpannableStringBuilder;

import com.dtu.capstone2.ereading.datasource.repository.EReadingRepository;
import com.dtu.capstone2.ereading.datasource.repository.LocalRepository;
import com.dtu.capstone2.ereading.network.request.DataStringReponse;
import com.dtu.capstone2.ereading.ui.model.LineContentNewFeed;
import com.dtu.capstone2.ereading.ui.model.TypeContent;

import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.Observable;
import io.reactivex.ObservableEmitter;
import io.reactivex.ObservableOnSubscribe;
import io.reactivex.Single;
import io.reactivex.functions.Function;

/**
 * Create by Nguyen Van Phuc on 4/9/19
 */
class TranslateNewFeedViewModel {
    private String urlNewFeed;
    private EReadingRepository mReadingRepository;
    private LocalRepository mLocalRepository;
    private List<String> mListVocabularyRefresh = new ArrayList<>();
    private SpannableStringBuilder mTextSpannableResultsContent = new SpannableStringBuilder();
    private SpannableStringBuilder mTextSpannableResultsTitle = new SpannableStringBuilder();
    private SpannableStringBuilder mTextSpannableResultsIntrudustion = new SpannableStringBuilder();

    TranslateNewFeedViewModel(EReadingRepository eReadingRepository, LocalRepository localRepository) {
        mReadingRepository = eReadingRepository;
        mLocalRepository = localRepository;
    }

    String getUrlNewFeed() {
        return urlNewFeed;
    }

    void setUrlNewFeed(String urlNewFeed) {
        this.urlNewFeed = urlNewFeed;
    }

    Observable<LineContentNewFeed> getDataFromHTMLAndOnNextDetectWord() {
        // TRường hợp với báo BBC text ok
        return Observable.create(new ObservableOnSubscribe<LineContentNewFeed>() {
            @Override
            public void subscribe(ObservableEmitter<LineContentNewFeed> emitter) throws Exception {
                Document doc = Jsoup.connect(urlNewFeed).get();
                emitter.onNext(new LineContentNewFeed(TypeContent.TITLE, doc.title(), null, null));

                Elements contents = doc.getElementsByTag("p");

                for (Element contentElem : contents) {
                    if (contentElem.hasClass(TypeContent.INTRODUCTION.getValueType())) {
                        emitter.onNext(new LineContentNewFeed(TypeContent.INTRODUCTION, contentElem.text(), null, null));
                    } else if (contentElem.hasClass(TypeContent.HEADER.getValueType())) {
                        emitter.onNext(new LineContentNewFeed(TypeContent.HEADER, contentElem.text(), null, null));
                    } else if (contentElem.hasClass(TypeContent.ITEM.getValueType())) {
                        emitter.onNext(new LineContentNewFeed(TypeContent.ITEM, contentElem.text(), null, null));
                    } else if (contentElem.attributes().size() == 0) {
                        emitter.onNext(new LineContentNewFeed(TypeContent.TEXT, contentElem.text(), null, null));
                    }
                }
                emitter.onComplete();
            }
        }).flatMapSingle(new Function<LineContentNewFeed, Single<LineContentNewFeed>>() {
            @Override
            public Single<LineContentNewFeed> apply(final LineContentNewFeed lineContentNewFeed) throws Exception {
                return mReadingRepository.GetDataStringReponse(lineContentNewFeed.getTextContent(), mLocalRepository.getNameLevelUser())
                        .map(new Function<DataStringReponse, LineContentNewFeed>() {
                            @Override
                            public LineContentNewFeed apply(DataStringReponse dataStringReponse) throws Exception {
                                return new LineContentNewFeed(lineContentNewFeed.getTypeContent(),
                                        dataStringReponse.getStringData(),
                                        dataStringReponse.getListVocabulary(),
                                        dataStringReponse.getListVocabularyNotTranslate());
                            }
                        });
            }
        });
    }

    void addOrRemoveVocabularyToListRefresh(String vocabulary) {
        vocabulary = vocabulary.toLowerCase();
        if (mListVocabularyRefresh.contains(vocabulary)) {
            mListVocabularyRefresh.remove(vocabulary);
        } else {
            mListVocabularyRefresh.add(vocabulary);
        }
    }

    int getSizeListRefresh() {
        return mListVocabularyRefresh.size();
    }

    public SpannableStringBuilder getmTextSpannableResultsContent() {
        return mTextSpannableResultsContent;
    }

    public SpannableStringBuilder getmTextSpannableResultsTitle() {
        return mTextSpannableResultsTitle;
    }

    public SpannableStringBuilder getmTextSpannableResultsIntrudustion() {
        return mTextSpannableResultsIntrudustion;
    }
}

File: app/src/main/java/com/dtu/capstone2/ereading/ui/newfeed/translate/TranslateNewFeedViewModel.kt
package com.dtu.capstone2.ereading.ui.newfeed.translate

import com.dtu.capstone2.ereading.datasource.repository.EReadingRepository
import com.dtu.capstone2.ereading.datasource.repository.LocalRepository
import com.dtu.capstone2.ereading.network.request.DataStringReponse
import com.dtu.capstone2.ereading.network.request.Vocabulary
import com.dtu.capstone2.ereading.network.response.DetailResponse
import com.dtu.capstone2.ereading.ui.model.LineContentNewFeed
import com.dtu.capstone2.ereading.ui.model.TypeContent
import com.dtu.capstone2.ereading.ui.model.VocabularyLocation
import com.dtu.capstone2.ereading.ui.model.VocabularySelected
import com.dtu.capstone2.ereading.ui.utils.RxBusTransport
import com.dtu.capstone2.ereading.ui.utils.Transport
import com.dtu.capstone2.ereading.ui.utils.TypeTransportBus
import io.reactivex.Observable
import io.reactivex.ObservableOnSubscribe
import io.reactivex.Single
import org.jsoup.Jsoup

/**
 * Create by Nguyen Van Phuc on 4/9/19
 */
internal class TranslateNewFeedViewModel(private val mReadingRepository: EReadingRepository, private val mLocalRepository: LocalRepository) {
    var urlNewFeed: String? = null
    private val mListVocabularyRefresh = mutableListOf<VocabularySelected>()
    private val mListVocabularyAddFavorite = mutableListOf<VocabularySelected>()
    private var mNameListDialogShowing: String = ""
    private val mListVocabularyTranslateResponse = mutableListOf<Vocabulary>()
    private val mListVocabularyNotTranslateResponse = mutableListOf<Vocabulary>()
    internal val dataRecyclerView = mutableListOf<LineContentNewFeed>()

    // TRường hợp với báo BBC text ok
    fun getDataFromHTMLAndOnNextDetectWord(): Observable<DataStringReponse> = Observable.create(ObservableOnSubscribe<LineContentNewFeed> { emitter ->
        with(Jsoup.connect(urlNewFeed).get()) {
            emitter.onNext(LineContentNewFeed(TypeContent.TITLE, this.title()))

            with(this.getElementsByTag("p")) contentElem@{
                this@contentElem.forEach { element ->
                    when {
                        element.hasClass(TypeContent.INTRODUCTION.valueType) -> emitter.onNext(LineContentNewFeed(TypeContent.INTRODUCTION, element.text()))
                        element.hasClass(TypeContent.HEADER.valueType) -> emitter.onNext(LineContentNewFeed(TypeContent.HEADER, element.text()))
                        element.hasClass(TypeContent.ITEM.valueType) -> emitter.onNext(LineContentNewFeed(TypeContent.ITEM, element.text()))
                        element.attributes().size() == 0 -> emitter.onNext(LineContentNewFeed(TypeContent.TEXT, element.text()))
                    }
                }
            }
        }
        emitter.onComplete()
    }).flatMapSingle { (typeContent, textContent) ->
        mReadingRepository.GetDataStringReponse(textContent, mLocalRepository.nameLevelUser)
                .doOnSuccess {
                    mListVocabularyTranslateResponse.addAll(it.listVocabulary)
                    mListVocabularyNotTranslateResponse.addAll(it.listVocabularyNotTranslate)
                    dataRecyclerView.add(LineContentNewFeed(typeContent,
                            it.stringData,
                            it.listVocabulary,
                            it.listVocabularyNotTranslate))
                }
    }

    fun getSizeListRefresh() = mListVocabularyRefresh.size

    fun getSizeListAddFavorite() = mListVocabularyAddFavorite.size

    fun addOrRemoveVocabularyToListRefresh(vocabularyLocation: VocabularyLocation) {
        mListVocabularyNotTranslateResponse.firstOrNull {
            it.startIndex == vocabularyLocation.startIndex && it.endIndex == vocabularyLocation.endIndex
        }?.let {
            with(VocabularySelected(vocabulary = it)) {
                mListVocabularyRefresh.firstOrNull { vocabularySelected ->
                    vocabularySelected.vocabulary.word == this.vocabulary.word
                }.let { vocabularySelected ->
                    when {
                        vocabularySelected == null -> {
                            mListVocabularyRefresh.add(this)
                        }
                        vocabularySelected.vocabulary.type != it.type -> {
                            mListVocabularyRefresh.add(this)
                        }
                        vocabularySelected.vocabulary.startIndex != vocabularyLocation.startIndex &&
                                vocabularySelected.vocabulary.endIndex != vocabularyLocation.endIndex -> {
                            RxBusTransport.publish(Transport(TypeTransportBus.TOAST_WITH_MESSAGE_SELECT_WORD,
                                    message = vocabularySelected.vocabulary.word))
                        }
                        else -> {
                            mListVocabularyRefresh.remove(this)
                        }
                    }
                }
            }
        }
    }

    fun addOrRemoveVocabularyToListAddFavoriteByLocationVocabulary(vocabularyLocation: VocabularyLocation) {
        mListVocabularyTranslateResponse.firstOrNull {
            it.startIndex == vocabularyLocation.startIndex && it.endIndex == vocabularyLocation.endIndex
        }?.let {
            with(VocabularySelected(vocabulary = it)) {
                mListVocabularyAddFavorite.firstOrNull { vocabularySelected ->
                    vocabularySelected.vocabulary.word == this.vocabulary.word
                }.let { vocabularySelected ->
                    when {
                        vocabularySelected == null -> {
                            mListVocabularyAddFavorite.add(this)
                        }
                        vocabularySelected.vocabulary.type != it.type -> {
                            mListVocabularyAddFavorite.add(this)
                        }
                        vocabularySelected.vocabulary.startIndex != vocabularyLocation.startIndex &&
                                vocabularySelected.vocabulary.endIndex != vocabularyLocation.endIndex -> {
                            RxBusTransport.publish(Transport(TypeTransportBus.TOAST_WITH_MESSAGE_SELECT_WORD, message = vocabularySelected.vocabulary.word))
                        }
                        else -> {
                            mListVocabularyAddFavorite.remove(this)
                        }
                    }
                }
            }
        }
    }

    fun getArrayWordRefresh() = mListVocabularyRefresh.map {
        it.vocabulary.word
    }.toTypedArray()

    fun getArrayWordAddFavorite() = mListVocabularyAddFavorite.map {
        it.vocabulary.word
    }.toTypedArray()

    fun getArraySelectedRefresh() = mListVocabularyRefresh.map {
        it.isChecked
    }.toBooleanArray()

    fun getArraySelectedAddFavorite() = mListVocabularyAddFavorite.map {
        it.isChecked
    }.toBooleanArray()

    fun resetListVocabularyRefresh() {
        mListVocabularyRefresh.clear()
    }

    fun resetListVocabularyAddFavorite() {
        mListVocabularyAddFavorite.clear()
    }

    fun getNameListDialogShowing() = mNameListDialogShowing

    fun setNameListDialogShowing(name: String) {
        mNameListDialogShowing = name
    }

    fun getPositionItemInsertedOfRV() = dataRecyclerView.size

    fun addFavoriteToServer(): Single<DetailResponse> = mReadingRepository.setListVocabularyFavorite(mListVocabularyAddFavorite.filter {
        it.isChecked
    }.map {
        it.vocabulary
    })
}
